aggregator.go

```diff
@@ -0,0 +1,300 @@
+	UniqueBlockCumulativeSize uint64 // optional dag size, used as the TSize in the unixfs link entry and recorded in manifest
+}
+
+type nodeMap map[string]*merkledag.ProtoNode
+
+// Aggregate de-duplicates and orders the supplied list of `AggregateDagEntry`-es
+// and adds them into a two-level UnixFSv1 directory structure. The intermediate
+// blocks comprising the directory tree and the manifest json file are written
+// to the supplied DAGService. No "temporary blocks" are produced in the process:
+// everything written to the DAGService is part of the final DAG capped by the
+// final `aggregateRoot`.
+func Aggregate(ds ipldformat.DAGService, toAggregate []AggregateDagEntry) (aggregateRoot cid.Cid, err error) {
+
+	dags := make(map[string]*ManifestDagEntry, len(toAggregate))
+
+	for _, d := range toAggregate {
+
+		cv1 := cid.Undef
+		if d.RootCid.Version() == 1 {
+			cv1 = d.RootCid
+		} else {
+			cv1 = cid.NewCidV1(d.RootCid.Type(), d.RootCid.Hash())
+		}
+
+		// do not consider identity multihashes for aggregation
+		if cv1.Prefix().MhType == uint64(multicodec.Identity) {
+			continue
+		}
+
+		cv0 := cid.Undef
+		if cv1.Prefix().Codec == uint64(multicodec.DagPb) &&
+			cv1.Prefix().MhType == uint64(multicodec.Sha2_256) &&
+			cv1.Prefix().MhLength == 32 {
+			cv0 = cid.NewCidV0(cv1.Hash())
+		}
+
+		dagName := cv1.String()
+
+		if _, exists := dags[dagName]; !exists {
+			dags[dagName] = &ManifestDagEntry{
+				rootCid:    cv1,
+				RecordType: DagAggregateEntry,
+				DagCidV1:   dagName,
+			}
+			if cv0 != cid.Undef {
+				dags[dagName].DagCidV0 = cv0.String()
+			}
+			if d.UniqueBlockCumulativeSize != 0 {
+				v := d.UniqueBlockCumulativeSize
+				dags[dagName].DagSize = &v
+			}
+			if d.UniqueBlockCount != 0 {
+				v := d.UniqueBlockCount
+				dags[dagName].NodeCount = &v
+			}
+		}
+	}
+
+	if len(dags) == 0 {
+		return cid.Undef, xerrors.New("no valid entries to aggregate")
+	}
+
+	// unixfs sorts internally, so we need to pre-sort to match up insertion indices
+	sortedDags := make([]*ManifestDagEntry, 0, len(dags))
+	for _, d := range dags {
+		sortedDags = append(sortedDags, d)
+	}
+	sort.Slice(sortedDags, func(i, j int) bool {
+		return strings.Compare(sortedDags[i].DagCidV1, sortedDags[j].DagCidV1) < 0
+	})
+
+	// innermost layer, 4 bytes off end
+	innerNodes := make(nodeMap)
+	for _, d := range sortedDags {
+
+		parentName := d.DagCidV1[:3] + `...` + d.DagCidV1[len(d.DagCidV1)-4:]
+		if _, exists := innerNodes[parentName]; !exists {
+			innerNodes[parentName] = emptyDir()
+		}
+
+		var dagSize uint64
+		if d.DagSize != nil {
+			dagSize = *d.DagSize
+		}
+
+		if err := innerNodes[parentName].AddRawLink(d.DagCidV1, &ipldformat.Link{
+			Size: dagSize,
+			Cid:  d.rootCid,
+		}); err != nil {
+			return cid.Undef, err
+		}
+		d.PathIndexes[2] = len(innerNodes[parentName].Links()) - 1
+	}
+
+	newBlocks := make([]ipldformat.Node, 0, 1<<20)
```

Comment:
Q: Use `len(innerNodes)` instead of `1 << 20`?

Reply:
Good point: though it also needs to accommodate the higher layer and the root. How about `len(innerNodes)*3/2` ?

Reply:
SGTM